src/util/taproot.rs

```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        self.serialize_to_writer(&mut buf)
+            .expect("writers don't error");
+        buf
+    }
+
+    /// Verify that a control block is correct proof for a given output key and script
+    /// This only checks that script is contained inside the taptree described by
+    /// output key, full verification must also execute the script with witness data
+    pub fn verify_taproot_commitment<C: secp256k1::Verification>(
+        &self,
+        secp: &Secp256k1<C>,
+        output_key: &schnorr::PublicKey,
+        script: &Script,
+    ) -> bool {
+        // compute the script hash
+        // Initially the curr_hash is the leaf hash
+        let mut curr_hash = compute_leaf_hash(&script, self.leaf_version);
+        // Verify the proof
+        for elem in self.merkle_branch.as_inner() {
+            let mut eng = TapBranchHash::engine();
+            if &curr_hash < elem {
+                eng.input(&curr_hash);
+                eng.input(elem);
+            } else {
+                eng.input(elem);
+                eng.input(&curr_hash);
+            }
+            // Recalculate the curr hash as parent hash
+            curr_hash = sha256::Hash::from_engine(eng);
+        }
+        // compute the taptweak
+        let tweak = compute_tap_tweak(Some(curr_hash), self.internal_key);
+        self.internal_key.tweak_add_check(
+            secp,
+            output_key,
+            self.output_key_parity,
+            tweak.into_inner(),
+        )
+    }
+}
+
+/// The leaf version for tapleafs
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LeafVersion {
+    ver: u8,
+}
+
+impl Default for LeafVersion {
+    fn default() -> Self {
+        Self {
+            ver: TAPROOT_LEAF_TAPSCRIPT,
+        }
+    }
+}
+
+impl LeafVersion {
+    /// Obtain LeafVersion from u8, will error when the last bit of ver is not 0
+    pub fn from_u8(ver: u8) -> Result<Self, TaprootError> {
+        if ver & TAPROOT_LEAF_MASK != ver {
+            Err(TaprootError::InvalidTaprootLeafVersion(ver))
+        } else {
+            Ok(LeafVersion { ver: ver })
+        }
+    }
+
+    /// Obtain a LeafVersion from u8 after applying TAPROOT_LEAF_MASK
+    pub fn from_unmasked_u8(ver: u8) -> Self {
+        LeafVersion { ver: ver & TAPROOT_LEAF_MASK }
+    }
+
+    /// Get the inner version from LeafVersion
+    pub fn as_u8(&self) -> u8 {
+        self.ver
+    }
+}
+
+impl Into<u8> for LeafVersion {
+    fn into(self) -> u8 {
+        self.ver
+    }
+}
+
+/// Detailed error type for taproot builder
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum TaprootBuilderError {
+    /// Merkle Tree depth must not be more than 128
+    InvalidMerkleTreeDepth(usize),
+    /// Nodes must be added specified in DFS order
+    NodeNotInDfsOrder,
+    /// Two nodes at depth 0 are not allowed
+    OverCompleteTree,
+    /// Invalid taproot internal key
+    InvalidInternalKey(secp256k1::Error),
```

Comment:
See my comment on `TaprootError` for similar variant there

Reply:
Answered this in another comment there. I think we are more likely to add the source if we keep the error type